rfcs/030-pipeline_merging/README.md

```diff
@@ -0,0 +1,124 @@
+sources: `transformerVersion*1000 + numberOfSources`.
+
+This means that 
+- a new `transformerVersion` always gets ingested
+- a work for the same `transformerVersion` and more sources gets ingested if there 
+are already works with the same `transformerVersion` but less sources
+
+##### Previous behaviour
+The ingestor used to calculate the version as `transformerVersion * 10 + (merged?1:0)`
+This meant that 
+- a new `transformerVersion` always got ingested
+- an out of order updates from the merger for the same `transformerVersion` might have been ingested 
+incorrectly cause they would have all the same elastic search version
+- unlink kind of worked because `merged` would have been set to `true` so, given 
+elastic search greater or equal than versioning, the unlinked version would have been 
+ingested (provided no message out of order issues)
+
+
+### Problems
+
+There are a number of issues with the current approach:
+- We need to multiply `transformerVersion` by 1000 to make it take precedence 
+over `numberOfSources` because `numberOfSources` can be quite high (650 is the recorded max).
+ If we don't do that we run into data consistency problems in the index. 
+ As we add more sources we will find ourselves incrementing this number regularly
+- The unliking case only works if the update that causes the unlink is on the target work,
+ therefore incrementing `transformerVersion`. If, as in the example above, the update is on a
+  work that gets redirected to the target, it won't be reflected in the API
+- The ingestor is responsible for figuring out the version based on information passed on
+ by the merger. This causes coupling between the merger and the ingestor. The ingestor is 
+ currently aware of merging happening at some point, which it shouldn't be. It also makes the 
+ versioning logic very hard to follow and modify
+
+### Discarded ideas
+
+Jamie and I came up with multiple ideas to tweak the current behaviour, each one with some problems:
+- Go back to the `merged` flag: it has issues with multiple sources and it has issues 
+with images versioning because image versions are derived from works
+- Add unlink in the version function like 
+`(transformerVersion * 10000) + (unlinked?1:0)*1000 + numberOfSources`: 
+bleurgh and also when something is ingested as unlinked it's impossible to override 
+it unless the `transformerVersion` changes
+- Use something like the sum of all `transformerVersion`s of all works sent by the matcher: 
+this would succeed in unlinking, but again would fail to add a work to a group 
+that was the result of a previous unlink operation
+
+
+### Solution proposed
+
+The solution suggested by Nic is that we pass through in the work a field with the timestamp 
+of the last update.
+The merger should then use the maximum of the timestamps of all works sent by the matcher 
+as the version for each of the works.
+The ingestor should use greater than versioning instead of greater or equal.
```

Comment:
Not sure I understand the reason for `gt` rather than `gte` :thinking:... I can't think of any data consistency issues that exist only with `gte` rather `gt`. One potential upside of `gte` would be that we could e.g. just redeploy the Calm transformer in an existing pipeline and run a Calm reindex where the data is retransformed (with `gt` any changes would be dropped).

Is there an issue I'm not thinking of? Sorry if we went over this in the meeting earlier.

Reply:
Mh yeah, good point about transformer changes, didn't think about that. I thought that there was consensus in the meeting that the way we use gte at the moment is dodgy and the new method wouldn't need it but I don't see how it would hurt, so yeah let's leave it